### PR TITLE
Detect out argument in the operator schema

### PR DESCRIPTION
**Pull request by tugsbayasgalan**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #64053
* #63651
* #63540
* #63096
* __->__ #62755

After this change, out argument can be checked by calling is_out(). We also update backward computability tests to make sure it works with both default and our arguments in the schema. 

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256/)

Differential Revision: [D30415256](https://our.internmc.facebook.com/intern/diff/D30415256)

### CODE REVIEW

**Review by iseeyuan — APPROVED** [review]
Looks good to me.

File: aten/src/ATen/core/function_schema.h [hunk 1]
```diff
@@ -34,6 +34,8 @@ struct Argument {
         default_value_(std::move(default_value)),
         kwarg_only_(kwarg_only),
         alias_info_(std::move(alias_info)) {
+    bool is_alias = alias_info_.has_value() && alias_info_.value().isWrite();
+    is_out_ = kwarg_only_ && is_alias;
```
Comment (iseeyuan) [inline aten/src/ATen/core/function_schema.h:38]: @gmagogsfm , could you confirm if this logic is good for all out args?

**Comment by iseeyuan** [comment]
@cccclai , you may consume is_out() here: https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/mobile/function.cpp#L100

File: test/cpp/jit/test_misc.cpp [hunk 2]
```diff
@@ -520,6 +520,14 @@ TEST(SchemaParserTest, NestedArrays) {
                                               .getElementType()));
 }
 
+TEST(SchemaParserTest, OutVariant) {
```
Comment (gmagogsfm) [inline test/cpp/jit/test_misc.cpp:523]: Could you add more test cases including:
1) operator without any out argument
2) `is_write` argument but not kwonly
3) arguments that are not `out` are correctly distinguished by `is_out`

**Review by gmagogsfm — CHANGES_REQUESTED** [review]

File: test/cpp/jit/test_misc.cpp (hunk 2 above)
Comment (tugsbayasgalan) [reply test/cpp/jit/test_misc.cpp:523]: will do thanks

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30415256).

**Review by raziel — COMMENTED** [review]
Curious if you think we should update the schema BC test function?

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30415256).

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30415256).

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30415256).

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30415256).

File: aten/src/ATen/core/function_schema_inl.h [hunk 3]
```diff
@@ -121,17 +121,40 @@ inline bool FunctionSchema::isBackwardCompatibleWith(
     }
   }
 
-  // Make sure that all the old arguments have their corresponding backward
-  // compatible arguments in this schema.
-  for (size_t i = 0; i < old.arguments().size(); ++i) {
+  // we want to test both out and default args seperately
+
+  // find the start of out args in the old schema
+  int old_out_start_idx;
```
Comment (gmagogsfm) [inline aten/src/ATen/core/function_schema_inl.h:127]: Future clean up:
It feels like that this loop has appeared often enough to warrant a utility method, what do you think?

File: aten/src/ATen/core/function_schema_inl.h [hunk 4]
```diff
@@ -121,17 +121,40 @@ inline bool FunctionSchema::isBackwardCompatibleWith(
...
+  for (old_out_start_idx = old.arguments().size() - 1; old_out_start_idx > -1;
```
Comment (gmagogsfm) [inline aten/src/ATen/core/function_schema_inl.h:128]:
```suggestion
  for (old_out_start_idx = old.arguments().size() - 1; old_out_start_idx >= 0;
```

File: aten/src/ATen/core/function_schema_inl.h [hunk 5]
```diff
@@ -121,17 +121,40 @@ inline bool FunctionSchema::isBackwardCompatibleWith(
...
+  for (size_t i = 0; i < old_out_start_idx; i++) {
     if (!arguments().at(i).isBackwardCompatibleWith(
           old.arguments().at(i), why_not)) {
       return false;
     }
   }
 
-  // Validate that all new arguments provided a default value.
-  for (size_t i = old.arguments().size(); i < arguments().size(); ++i) {
+  // Validate that all new arguments provided a default value
```
Comment (gmagogsfm) [inline aten/src/ATen/core/function_schema_inl.h:156]:
```suggestion
  // Validate that all new arguments provided has a default value
```

**Review by gmagogsfm — APPROVED** [review]

File: aten/src/ATen/core/function_schema_inl.h (hunk 5 above)
Comment (tugsbayasgalan) [reply aten/src/ATen/core/function_schema_inl.h:156]: thanks will do

File: aten/src/ATen/core/function_schema_inl.h (hunk 4 above)
Comment (tugsbayasgalan) [reply aten/src/ATen/core/function_schema_inl.h:128]: will do

File: aten/src/ATen/core/function_schema_inl.h (hunk 3 above)
Comment (tugsbayasgalan) [reply aten/src/ATen/core/function_schema_inl.h:127]: will di

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30415256).

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30415256).

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30415256).

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30415256).

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30415256).

**Pull request closed without merging**